src/webgl/shaders/imageLightDiffused.frag

```diff
@@ -0,0 +1,82 @@
+precision mediump float;
+varying vec3 localPos;
+
+// the HDR cubemap converted (can be from an equirectangular environment map.)
+uniform sampler2D environmentMap;
+varying vec2 vTexCoord;
+
+const float PI = 3.14159265359;
+
+// not needed
+vec2 normalToEquirectangular( vec3 v){
+  vec2 uv;
+  // taking the arctangent of the v.z and v.x components and dividing it by the circumference of a circle
+  // Adding 0.5 ensures that the resulting value is in the range [0, 1].
+  uv.x = atan(v.z, v.x) / 3.14159 + 0.5;
+  // taking the arcsine of the v.y component and dividing it by the half circumference of a circle
+  uv.y = asin(v.y) / 3.14159 + 0.5;
+  return uv;
+}
+
+vec2 nTOE( vec3 v ){
+  // x = r sin(phi) cos(theta)   
+  // y = r cos(phi)  
+  // z = r sin(phi) sin(theta)
+  float phi = acos( v.y );
+  // if phi is 0, then there are no x, z components
+  float theta = 0.0;
+  // else 
+  theta = acos(v.x / sin(phi));
+  float sinTheta = v.z / sin(phi);
+  if (sinTheta < 0.0) {
+    // Turn it into -theta, but in the 0-2PI range
+    theta = 2.0 * PI - theta;
+  }
+  theta = theta / (2.0 * 3.14159);
+  phi = phi / 3.14159 ;
+  
+  vec2 angles = vec2( phi, theta );
+  return angles;
+}
+
+void main()
+{   	 
+	// the sample direction equals the hemisphere's orientation
+  float phi = vTexCoord.x * 2.0 * PI;
+  float theta = vTexCoord.y * PI;
+  float x = sin(theta) * cos(phi);
+  float y = sin(theta) * sin(phi);
+  float z = cos(theta);
+  vec3 normal = vec3( x, y, z);
+
+	// Discretely sampling the hemisphere given the integral's spherical coordinates translates to the following fragment code:
+	vec3 irradiance = vec3(0.0);  
+	vec3 up	= vec3(0.0, 1.0, 0.0);
+	vec3 right = normalize(cross(up, normal));
+	up = normalize(cross(normal, right));
+
+	//  We specify a fixed sampleDelta delta value to traverse the hemisphere; decreasing or increasing the sample delta will increase or decrease the accuracy respectively.
+	const float sampleDelta = 0.025;
+	float nrSamples = 0.0;
+  
+	for(float phi = 0.0; phi < 2.0 * PI; phi += sampleDelta)
```

Comment:
Btw, I think the indentation looks off on Github because this line uses tabs and the next line uses spaces. Maybe vscode has as button to convert everything to spaces?

Reply:
I'll manually do this, vscode has a feature for this but it's messes up the whole file. It's kinda like prettier, the way prettier messes up more than it cleans :")